### DISCUSSION TITLE
Replacing indirect objects

### DISCUSSION THREAD

**Discussion by silicontrip** [qpdf/qpdf#507] [Q&A]
I'm having a little bit of trouble with `QPDF::replaceObject(objid,gen,pdfHandle)` If I use `QPDF::getObjectByID(objid,gen)` after the replacement the correct object is returned.  However using `QPDFWriter` is returning a pdf with the old object. 
I'm using QPDFWriter to get an in memory representation of the PDF, which I can pass to the Cocoa PDFView, (while editing a page content stream this gives real time feedback on my edits, it's quite fun to change the set colour or character spacing and watch the whole document change instantly) If there is a better way to get a buffer of PDF, it would be good to hear about it.
```
	QPDFWriter qpdfWriter(qdoc);
   // qpdfWriter.setOutputFilename("before.pdf");
	qpdfWriter.setOutputMemory();
	qpdfWriter.write();
	
	Buffer* qBuf = qpdfWriter.getBuffer();
	unsigned char const* qBytes = qBuf->getBuffer();

	CFStringRef cs = CFStringCreateWithBytes(NULL, qBytes, qBuf->getSize(), 0, false);

	QPDFObjectHandle qObject = QPDFObjectHandle::parse("<< /test 2 /Count 0 /Kids [ ] /Type /Pages >>");
	qdoc.replaceObject(2,0,qObject);
```
full code is here; https://gist.github.com/silicontrip/b2bef4cc3ca8cd4ebcc83487a8c3b344 

I did read `create_pdf.cc` and I think I might be going about creating a PDF the wrong way.  In the add_page function, an entire page is build in a seperate pdfobjecthandle from the PDF and then added to the PDF last.  I don't quite have the luxury of creating a seperate object tree, as my UI is a view of the PDF, everything that is shown is required to be in the document somewhere.   I need to find a way to add a "blank" object into the pdf structure and editing that.  Hence why I'm looking at `replaceObject`

**Comment by jberkenbilt** [comment]
Sorry for the delay. There's no expectation of being able to write a QPDF object with more than one QPDFWriter. It's not a configuration I test or support since QPDFWriter makes changes to the QPDF object in the writing process. There are apparently also things you have to understand about interoperation between C++ and Objective C/Objective C++ that I know nothing at all about but that another user ran into issues with in #502. If you call replaceObject before the first write and remove the second write, does the first write contain the updated information? Not that long ago, I attempted to update the code so that I could support the pattern of creating more than one writer for the same QPDF, but it would require pretty deep and risky changes to the library, so I haven't pursued it.

**Reply by silicontrip** [reply]
Yes, that works.  I can easily make the change in my code to only having one `QPDFWriter`.  But my more general question would be, does QPDF support this pattern; making a change, then writing (or getting PDF in memory), then making another change, then writing again, change, write, change and so on?   If not this is a bit of a show stopper for me, as the macOS PDF class requires initialisation from PDF data for its onscreen view.   So far I've only been using `replaceKey`, `setArrayItem` and `replaceStreamData` on the PDF and only encountering such issues as I'm adding features.

**Reply by jberkenbilt** [reply]
qpdf does not support that pattern, but you can always reread the file you have just written. With the exception of the ID, qpdf is idempotent, so if you create a QPDF, write, create a new QPDF from what you just wrote, and wrote again, the result will be unchanged except for the ID. If you want even the ID to match, use `writer.setDeterministicID(true)`. That pattern is heavily tested. If you add new objects, the object numbering may change, so you may have to retraverse stuff. Regarding efficiency, there will be no real performance difference in creating a new QPDF each time because QPDF loads lazily from the input file, and every write does a full traversal. So, in other words, creating a new QPDF object on the file you just written after each write will have essentially the same performance and the same output as what you are trying to do with multiple writes from the same file. The only difference is that whatever internal state you have regarding where things are in the QPDF will need to be updated. Granted, that will be annoying, but hopefully it's something you can manage. The `getRenumberedObjGen` method in `QPDFWriter` can help you here since it will give enable you to map from your old objects to new ones.

If you want, I can throw together a very short example if my explanation is not adequate.

**Reply by jberkenbilt** [reply]
Note that if you request an example, it may take me a couple of days. I've got a very full day tomorrow during which I will be offline most of the day. I may have some time waiting around with my laptop when I can throw something together, but it may be Thursday before I have time to actually do it. But I have no issue doing it if it would help. You can even use processMemoryFile on the result of the buffer you get back from setOutputMemory; just be sure to manage things so you're not leaking memory. I know this won't be as convenient, but I think it can be made to work pretty well.

**Reply by silicontrip** [reply]
All good, thanks again, this was a pretty trivial change, requiring 3 extra lines.  I'll see what the performance is like.
```
-(NSData*)data
{
	QPDFWriter qwriter(*qDocument);  //qDocument is a QPDF* instance variable
	qwriter.setOutputMemory();
	qwriter.write();
	
	Buffer* qBuf = qwriter.getBuffer();
	unsigned char const* qBytes = qBuf->getBuffer();
	
	NSData* pdfData = [[[NSData alloc] initWithBytes:qBytes length:qBuf->getSize()] autorelease];  //initWithBytes makes a copy of the buffer.
	delete(qBuf);

/* new code start */
	delete(qDocument);	
	qDocument = new QPDF(); 
	qDocument->processMemoryFile("NSData", (char*)[pdfData bytes], [pdfData length]); 
/* new code end */

	return pdfData;
}
```

**Reply by silicontrip** [reply]
Actually I'm not going to be able to make that work.  As the UI classes store references to my wrapper class for `QPDFObjectHandle` by invalidating these this would reset the document view state and be a nightmare for usability.  Trying to relink the UI elements to the new PDF, doesn't sound too easy.   Maybe I could make a deep copy of the QPDF document, and then perform a write on the copy.  So I can still edit the original document and the PDFView has the copy.
This would be one of the feature requests, being able to get the current representation of the PDF.  Or even just a single specified page as a PDF as I noticed that the updates were a bit "laggy" for very large PDF documents.

**Reply by silicontrip** [reply]
My ideal workflow would be to make a new PDF document based on the current page being viewed.  I'm assuming it'd be easy to have a base document template which could have the page dictionary grafted onto it.   This way the pdf view could be updated much faster, as I have tried to edit a large document and it was noticeably slower.  Only needing to generate the current page and not the whole document would probably save on resources. eg `QPDF* singlePageDocument = workingDocument->documentForPage(0);`  which would copy all the children of the page object.  I think I've seen something similar to this in the headers somewhere.

**Comment by jberkenbilt** [comment]
Feel free to create an issue asking for this enhancement. I made a note to myself to look at making it work. I'm guessing it will be doable, and honestly I've sometimes wanted to use this pattern myself to write two different versions of a file. My goal will be to make it possible to write the same QPDF object with different writers, possibly with intervening changes, not to reuse a QPDFWriter object, which I think is harder. Feel free to open a really simple issue just asking for the ability to write the same QPDF object through multiple QPDFWriters wit intervening changes and reference this discussion. I think it would be a very useful enhancement. I'll look at the code soon to try to assess how hard it is.

**Comment by jberkenbilt** [comment]
Ya know, I was looking through the code, and there's actually very little that precludes writing multiple times. In fact, I tried this little test program, and it works as expected. I wonder if something else is going on with your code.
```c++
#include <qpdf/QPDF.hh>
#include <qpdf/QPDFObjectHandle.hh>
#include <qpdf/QPDFWriter.hh>

int main()
{
    QPDF q;
    q.processFile("/tmp/a.pdf");
    QPDFWriter w1(q, "/tmp/1.pdf");
    w1.write();
    auto pages = q.getAllPages();
    auto p1 = pages.at(1);
    p1.replaceKey("/X", QPDFObjectHandle::newInteger(1));
    QPDFWriter w2(q, "/tmp/2.pdf");
    w2.write();
    p1.replaceKey("/X", QPDFObjectHandle::newInteger(2));
    QPDFWriter w3(q, "/tmp/3.pdf");
    w3.setQDFMode(true);
    w3.write();
    return 0;
}
```

**Reply by silicontrip** [reply]
Yes, I encountered that also,  editing streams, dictionaries and arrays by replacing keys or indexes works without issue,  my code even made a new `QPDFWriter` each time.  There were a lot of these calls, as each time the stream editor made a single character change as I was typing, a new PDF was generated each time.  Only when I started working with creating, deleting and replacing indirect objects did I encounter the issue using; `QPDF::replaceObject(int,int,QPDFObjectHandle)` I did try your code, however I don't have a correctly created `a.pdf`

**Comment by jberkenbilt** [comment]
Can you give me a really streamlined example of what you're trying to? I can create an example that does it or else figure out why it doesn't work and, if it's simple, fix it, or otherwise make a note or suggest an alternative. I had forgotten but I remember now having gone through and fixed most of the places where QPDFWriter mutates the underlying object when supporting custom stream filters back in December. Funny how fast we can forget.

**Reply by silicontrip** [reply]
sorry for the rushed response.
```
#include <qpdf/QPDF.hh>
#include <qpdf/QPDFObjectHandle.hh>
#include <qpdf/QPDFWriter.hh>

int main (int argc, char* argv[])
{

	QPDF qdoc;
	qdoc.emptyPDF();

	QPDFObjectHandle qObject = QPDFObjectHandle::parse("<< /test 2 /Count 0 /Kids [ ] /Type /Pages >>");

	QPDFWriter qpdfWriter(qdoc,"./before.pdf");
	qpdfWriter.write();
	
	qdoc.replaceObject(2,0,qObject);  // THIS ONE!!!



	QPDFWriter qpdf2Writer(qdoc,"./after.pdf");
	qpdfWriter.write();

	return 0;
}

```

**Reply by jberkenbilt** [reply]
Yeah, this crashes for me, and that should never happen. You have definitely uncovered a bug. But why not just add `/test 2` to the existing object? Why use replaceObject at all? Granted, it should work, and I think I will try to fix this for qpdf 11, but it would be better if you mutate objects without replacing them. Is there a deep reason why you have to replace objects, like you are changing their type? Rather than replacing arrays and dictionaries, you should just remove items from them. You can't mutate scalars, but indirect scalars are a little of a pathological case.

But even my suggested modification still crashes, so there's definitely something wrong.

```c++
#include <qpdf/QPDF.hh>
#include <qpdf/QPDFObjectHandle.hh>
#include <qpdf/QPDFWriter.hh>

int main (int argc, char* argv[])
{

    QPDF qdoc;
    qdoc.emptyPDF();

#if 0
    QPDFObjectHandle qObject = QPDFObjectHandle::parse("<< /test 2 /Count 0 /Kids [ ] /Type /Pages >>");
#else
    QPDFObjectHandle qObject = qdoc.getObjectByObjGen(QPDFObjGen(2, 0));
#endif

    QPDFWriter qpdfWriter(qdoc,"./before.pdf");
    qpdfWriter.write();

#if 0
    qdoc.replaceObject(2,0,qObject);  // THIS ONE!!!
#else
    qObject.replaceKey("/test", QPDFObjectHandle::newInteger(2));
#endif


    QPDFWriter qpdf2Writer(qdoc,"./after.pdf");
    qpdfWriter.write();

    return 0;
}
```

It's a rare event that someone draws to my attention something that causes qpdf to crash. Granted you are using the library in a way that is not tested or supported, but I can definitely see that it would be nice if it worked and that the current behavior is quite counterintuitive. And qpdf should not segfault. I'll give this some thought. I don't have a quick fix for you, I'm afraid.

**Comment by jberkenbilt** [comment]
Note that, when you call replaceObject, existing QPDFObjectHandles will still point to the old object. The comments in QPDF.hh explain this.
There's really nothing that can be done about that.

**Comment by jberkenbilt** [comment]
You could probably put a wrapper around QPDFObjectHandle that hangs onto some kind of indicator of what revision of your state it comes from...so when you replace an object, increment some global counter, and when create a wrapper object, store its current value. When you retrieve the object from your wrapper, if the global counter has changed, replace it with `qpdf.getObjectByObjGen(old_obj.getObjGen())`. That may be all you need to make this work.

**Comment by jberkenbilt** [comment]
Yeah, the issue here is with replaceObject. This little program illustrates how it behaves.

```c++
#include <qpdf/QPDF.hh>
#include <qpdf/QPDFObjectHandle.hh>
#include <qpdf/QPDFWriter.hh>
#include <iostream>
#include <cassert>

class ObjWrapper
{
  public:
    ObjWrapper(int state, QPDF& q, QPDFObjectHandle obj) :
        state(state),
        q(q),
        obj(obj)
    {
        if (! obj.isIndirect())
        {
            throw std::runtime_error("you can only wrap indirect objects");
        }
    }

    QPDFObjectHandle getObj(int state)
    {
        if (this->state != state)
        {
            this->state = state;
            this->obj = q.getObjectByObjGen(this->obj.getObjGen());
        }
        return this->obj;
    }

  private:
    int state;
    QPDF& q;
    QPDFObjectHandle obj;
};

int main()
{
    static int state = 0;
    QPDF q;
    q.processFile("/tmp/a.pdf");
    QPDFObjectHandle orig_thing = q.makeIndirectObject(
        QPDFObjectHandle::newInteger(1));
    assert(orig_thing.getIntValue() == 1);
    QPDFObjectHandle orig_root = q.getRoot();
    orig_root.replaceKey("/Thing", orig_thing);
    ObjWrapper root_wrapper(state, q, orig_root);
    ObjWrapper thing_wrapper(state, q, orig_thing);
    assert(thing_wrapper.getObj(state).getIntValue() == 1);

    // Now replace the object and increment state.
    QPDFObjectHandle new_value = QPDFObjectHandle::newInteger(2);
    q.replaceObject(orig_thing.getObjGen(), new_value);
    ++state;

    // This one is not updated.
    assert(orig_thing.getIntValue() == 1);
    // Also, the old root already contains a resolved object
    assert(orig_root.getKey("/Thing").getIntValue() == 1);
    // This gets the new value.
    assert(thing_wrapper.getObj(state).getIntValue() == 2);

    // Unfortunately, the cached root's /Thing points to the old object.
    assert(root_wrapper.getObj(state).getKey("/Thing").getIntValue() == 1);

    // We can fix it, but we have to actually do that.
    orig_root.replaceKey("/Thing", thing_wrapper.getObj(state));
    // Now we have the new value.
    assert(root_wrapper.getObj(state).getKey("/Thing").getIntValue() == 2);
    return 0;
}
```

The thing is that, internally, a QPDFObjectHandle contains a smart pointer to the underlying object. There are no backwards reference to where other copies of that object might exist. When you create a QPDFObjectHandle, the object pointer is initially null, and when you first access it, it goes and fills it in. Dictionaries are just STL maps of string keys to QPDFObjectHandles. So traversing through an object will cause the cached copy of that object to internally contain resolved objects, and those will continue to point to old objects.

In retrospect, this is a bad design, and I wouldn't be surprised if, now having thought this through freshly, I couldn't actually think of cases where this would case deeper confusion. I might be able to fix it, but almost surely not without breaking binary compatibility. I think I would actually have to store something in the underlying object that indicates that it has been replaced, which would in turn force all attempts to access it to re-resolve it. It would be a small fix, I think, but it would require changing the internals of QPDFObject, which breaks binary compatibility.

But I think this will have to be fixed because I don't think there's any hope of ever being able to keep track of all the places where an object might have been resolved in the general case.

**Reply by silicontrip** [reply]
Sounds like a design implementation problem. Ouch!  I should be able to work around needing to use `replaceObject`  It was really for completeness that I had add/remove buttons on all 3 outline views.  The 3 views simply came about because I thought it would be nice to present the PDF in different ways.  There's nothing stopping me from only using the Trailer View to add, edit or remove.  And it sounds like I'd have much better luck with just replacing dictionary keys or array indexes.  My current goal is to be able to have enough functionality to create a text pdf from an empty PDF (currently I have a test document with 1 blank page to experiment drawing on it)

**Reply by jberkenbilt** [reply]
In spite of my fixing the replaceObject issue, I still think just replacing dictionary keys and array indexes will be a better approach. Also, there is a bit of a performance penalty across the whole library from this fix. I've coded it so the extra check only happens once relpaceObject (or swapObjects, which has the same issue) is called for the first time. If you never call replaceObject or swapObjects, there is a 1% to 2% hit in my performance test, which I think is worth it for eliminating this problem. Once you have called replaceObject or swapObjects, the hit jumps up to around 7% in the worse case (writing linearized files) but floats at around 2% to 5%. This is very sad, but I also think this is an important fix. Hopefully I won't have to back it out for unintended consequences. Technically it is a behavior change that could change the meaning of previously non-crashing code, but it's almost unthinkable to me that someone would be actually relying on the old behavior.

**Comment by jberkenbilt** [comment]
It's also worth noting that there are very few places within qpdf that internally use replaceObject, and it's mostly in cases where there are no opportunities for the object to have been dereferenced by a user, but I still think there is a potential source of a bug here.

**Reply by jberkenbilt** [reply]
worth noting, not worth nothing...

**Comment by jberkenbilt** [comment]
I realized overnight that there's actually a very easy fix to this problem that doesn't break binary compatibility at all. When QPDFObjectHandle goes to dereference the underlying object, it just makes a quick check to see if the object has changed and replaces it it if it has. After this very small change, old QPDFObjectHandles automatically update after calling replaceObject. I will make this fix for 10.2.1 which I will try to get out in the next few days.

I will also investigate the write() segfault.

**Reply by jberkenbilt** [reply]
Heh, the segfault for write is just a typo in your code. You call `qpdfWriter.write()` instead of `qpdf2Writer.write()`. Took me running gdb to notice the typo. :-) With my current fix and the typo fix, your sample code works as you expect and as it should: the /test 2 is visible in after.pdf.

So hang tight a few days for 10.2.1, and if you want a head start, I'll be pushing a fix to master soon. I need to update docs and let it pass CI, which takes 30 to 40 minutes. I'll post here when the fix is in master.

Wow, this has been lurking in qpdf since August, 2011, and the fix was so small. I'm really glad you brought this up. I feel a lot better about the behavior of replaceObject after this fix.

Also, I misspoke -- I do actually test the case of writing the same QPDF through multiple writers, so that should work fine. In other words, after 10.2.1 is out, you should be able to proceed with your original design. If not, please let me know, and I will continue to investigate.

**Comment by jberkenbilt** [comment]
This code illustrates how 10.2.1 will behave.

```c++
#include <qpdf/QPDF.hh>
#include <qpdf/QPDFObjectHandle.hh>
#include <iostream>

int main()
{
    QPDF q;
    q.emptyPDF();
    QPDFObjectHandle orig_thing = q.makeIndirectObject(
        QPDFObjectHandle::newInteger(1));
    assert(orig_thing.getIntValue() == 1);
    QPDFObjectHandle orig_root = q.getRoot();
    orig_root.replaceKey("/Thing", orig_thing);

    QPDFObjectHandle new_value = QPDFObjectHandle::newInteger(2);
    q.replaceObject(orig_thing.getObjGen(), new_value);

    // This one is updated now.
    assert(orig_thing.getIntValue() == 2);
    // This one too.
    assert(orig_root.getKey("/Thing").getIntValue() == 2);

    return 0;
}

```

**Reply by jberkenbilt** [reply]
(edited to remove unused `state` variable)

**Comment by jberkenbilt** [comment]
I'm writing up a quick summary here since the answer is spread out across a complex chain of replies.

* 10.2.0 behavior and earlier: after `replaceObject()` is called, existing `QPDFObjectHandle` instances that have been previously resolved will continue to point to the old object. Those that have reference to the object and have not been resolved will point to the new object when resolved. The resolution process happens internally to the library the first time a value is queried. This is clearly undesirable behavior.
* 10.2.1 behavior: existing `QPDFObjectHandle` instances for indirect objects will notice when the underlying object changed and update themselves. This happens at a slight performance cost of 1% to 2% in code that never calls `replaceObject()` and more like 5% to 7% in code that does. While unfortunate, it seems like this is an important fix.

The originally suggested workflow of using `replaceObject()` should probably work from 10.2.1, but just mutating dictionaries and arrays is a more robust way to implement it and will also perform better.

Additionally, I misspoke about not supporting using multiple `QPDFWriter`s to write the same `QPDF` multiple times. This is actually exercised in the test suite, and it should be safe to do that. If there are problems with that, a separate issue should be reported. The crash reported in this original description was actually because of a typo in the variable name in the code, not because of a crash in the qpdf library.

**Reply by jberkenbilt** [reply]
Fixed in a4d6589ff26007c966db8912e4dae1aa937a5968

**Comment by jberkenbilt** [comment]
@silicontrip In case you're not subscribed to qpdf-announce, 10.3.0 is out and includes the fix for this. Feel free to give it another try, unless you're happy with your (probably better) solution of not using relpaceObject.

**Reply by silicontrip** [reply]
Hi Jay,
I've put editing the object view on hold for the moment as I've been away trying to straighten out my UI code which was starting to turn into spaghetti.  Anyway I'm wondering how to create an indirect object, I discovered that newIndirect and objGenToIndirect are private.  My specific example (though there are probably others) is to add the `/Parent` indirect object in a `/Page` dictionary.  But I'm unable to create the needed object.
Thanks

**Reply by jberkenbilt** [reply]
QPDF::makeIndirectObject

**Reply by jberkenbilt** [reply]
Though I'm not sure why you would create a new indirect object to add `/Parent` to a page dictionary. You would pretty much already have to have a QPDFObjectHandle for the parent. Also, there are interfaces in QPDFPageDocumentHelper for adding pages without having to manipulate the page dictionary directly. I'm going to be adding new interfaces for that soon.

**Reply by silicontrip** [reply]
I didn't mean, make a new indirect object, I mean how do I parse the string '2 0 R' so that object becomes an indirect object.  So it sounds like I need to determine the object ID and generation from the string, then get the object by ID, and replace that as the element, in either a dictionary or array.  I'll see how I go.
For the moment I'm finishing all the basic functionality, when I can hand roll an entire PDF, then I'll look at higher level tasks.  Although I'll probably be looking for insertFont or insertBitmapImage before then.

**Reply by jberkenbilt** [reply]
You can use the new version of parse introduced in 10.2.0: `QPDFObjectHandle::parse(qpdf, "2 0 R")`, which will be interpreted as that indirect object in the given QPDF. You could also call `QPDF::getObjectById` if you parsed the string yourself. That would work in older code as well. It's puzzling to me why you would need to do this. How are you getting the string? qpdf would never give you that string unless you called unparse() on an object, in which case you already have the object.

**Reply by jberkenbilt** [reply]
Actually, I think literally calling parse on "2 0 R" would fail because it would just read the 2. You can fake it out by calling `QPDFObjectHandle::parse(qpdf, "[2 0 R]").getArrayItem(0)`.

**Reply by silicontrip** [reply]
OMG!!! it works! I have just created the first PDF from scratch simply by adding the needed objects.  Not the most exciting PDF but shows that my idea is feasible.
<img width="976" alt="20210322-NewPageScaled" src="https://user-images.githubusercontent.com/1704192/111912717-01e36b00-8abf-11eb-93b0-e2e95a5af379.png">

**Reply by silicontrip** [reply]
I'm actually using 10.3.1 now

**Reply by jberkenbilt** [reply]
Being the first one created from scratch in this way by your application seems like enough to make it an exciting PDF. :-)

**Reply by silicontrip** [reply]
I've been a bit busy with the Syntax highlighter, if you're interested 
<img width="640" alt="SS210410" src="https://user-images.githubusercontent.com/1704192/114265157-a3ad0680-9a32-11eb-9951-6eeeb34915a1.png">
